tests/Unit/PackagePrivate/InMemoryTermIdsAcquirerTest.php

```diff
@@ -91,4 +95,78 @@ private function assertNoDuplicates( $ids ) {
 		);
 	}
 
+	public function testCleanTerms_doesNotReuseIds() {
+		$termIdsAcquirer = new InMemoryTermIdsAcquirer();
+
+		$ids1 = $termIdsAcquirer->acquireTermIds( [
+			'label' => [
+				'en' => 'the label',
+				'de' => 'die Bezeichnung',
+			],
+			'alias' => [
+				'en' => [ 'alias', 'another' ],
+			],
+		] );
+		$ids2 = $termIdsAcquirer->acquireTermIds( [
+			'label' => [ 'en' => 'the label' ],
+			'description' => [ 'en' => 'the description' ],
+		] );
+
+		$termIdsAcquirer->cleanTerms( array_merge( $ids1, $ids2 ) );
+
+		$ids3 = $termIdsAcquirer->acquireTermIds( [
+			'label' => [ 'en' => 'the label' ],
+			'description' => [ 'en' => 'the description' ],
+		] );
+		$this->assertGreaterThan( max( max( $ids1 ), max( $ids2 ) ), min( $ids3 ) );
+	}
+
+	public function testCleanTerms_completelyCleansArray() {
+		$termIdsAcquirer = new InMemoryTermIdsAcquirer();
+
+		$ids1 = $termIdsAcquirer->acquireTermIds( [
+			'label' => [
+				'en' => 'the label',
+				'de' => 'die Bezeichnung',
+			],
+			'alias' => [
+				'en' => [ 'alias', 'another' ],
+			],
+		] );
+		$ids2 = $termIdsAcquirer->acquireTermIds( [
+			'label' => [ 'en' => 'the label' ],
+			'description' => [ 'en' => 'the description' ],
+		] );
+
+		$termIdsAcquirer->cleanTerms( array_merge( $ids1, $ids2 ) );
+
+		$this->assertEmpty( TestingAccessWrapper::newFromObject( $termIdsAcquirer )->terms );
+	}
+
+	public function testCleanTerms_keepsOtherIds() {
+		$termIdsAcquirer = new InMemoryTermIdsAcquirer();
+
+		$ids1 = $termIdsAcquirer->acquireTermIds( [
+			'label' => [
+				'en' => 'the label',
+				'de' => 'die Bezeichnung',
+			],
+			'alias' => [
+				'en' => [ 'alias', 'another' ],
+			],
+		] );
+		$ids2 = $termIdsAcquirer->acquireTermIds( [
+			'label' => [ 'en' => 'the label' ],
+			'description' => [ 'en' => 'the description' ],
+		] );
+
+		$termIdsAcquirer->cleanTerms( array_diff( $ids1, $ids2 ) );
+
+		$ids3 = $termIdsAcquirer->acquireTermIds( [
+			'label' => [ 'en' => 'the label' ],
+			'description' => [ 'en' => 'the description' ],
+		] );
+		$this->assertSame( $ids2, $ids3 );
```

Comment:
Another follow up: https://github.com/wmde/mediawiki-term-store/pull/19

This test is not doing what I originally expected it to be doing: ensuring only unreferenced ids are cleaned up. The tested code (InMemoryTermIdsAcquirer) does not seem to be doing that at all. Should it not be doing that?

Reply:
How is the code supposed to know which term_in_lang IDs are referenced and which aren’t? It cleans up all the ones it gets, that’s why the calling code has to do the `array_diff` (`$ids2` are still used).

As far as I can tell, your follow-up is mostly equivalent, except that it hard-codes all the IDs, meaning it will break as soon as we change the ID assignment in any way (e. g. start counting from 0 instead of 1, or make the counter per-class instead of per-instance), so I don’t understand how it’s supposed to be an improvement.

Reply:
The refactored version is less complex and (at least to me) easier to understand. If the array_diff was adding some kind of value to the test, it is unclear enough for me to still not understand it after looking at it for 10 minutes.

You are right this test binds to the id numbers. Not clear to me there is a practical downside to that for this particular case. Remember we are talking about an in memory test double that is package private.

Reply:
The non-in-memory implementation is also “package private” at the moment, so to me that label doesn’t carry much value. Plenty of Wikibase’s in-memory test doubles are public (because they’re useful in tests of other components as well).

Reply:
Yes many test doubles in WB are public and used in other components. (A practice which I started, and had some people complaining about consistency and "this should not be public because I don't expect it to be".)

So what? And how does that make the distinction between package public and private less important? Does it not make it more important?

Reply:
re PackagePrivate .. in order to proceed with that discussion we have to catch up on what Lucas and I have decided (but haven't encoded in this library's design yet) first.

In short, we decided that this library should be concerned with only the tabled `wbt_term_in_lang`, `wbt_text_in_lang`, `wbt_type` and `wbt_text`. Hence the interfaces TermIdsAcquirer and TermCleaner are to be public and the main interface of this library.

Since Term is a concept in Wikibase Data Model, then modeling this library as to abstract normalized stored terms starting from terms (and not be concerned about linking them to entities) is okay conceptually, and would make this design more independent of what entity types might come in and use it.

`wbt_item_terms` and `wbt_property_terms` are tables that link specific entity types to terms. Other entity types might be defined later and link to terms in this store in different ways and that "term-store" should not be concerned of that to make it simpler.

So the way to move forward is:
* drop PropertyTermStore and ItemTermStore interfaces from wikibase-term-store library
* move TermIdsAcquirer, TermsCleaner and their InMemory implementations into wikibase-term-store library that provides the design of our term store solution.
* this library provides implementation details that depend on mediawiki dbal and speicifics (like replica/master related logic)
* Wikibase extension is then to worry about handling the tables `wbt_item_terms` or `wbt_property_terms` (or whatever that suits its needs while linking entities to terms), and other extensions will then also do the same in their own scope.

Reply:
This makes no sense to me and I have commented in the past that this is what I think about such an approach.

Reply:
A bunch of thoughts on this topic:

* From the perspective of the "store Item/Property terms" use case
    * ItemTermStore and PropertyTermStore are good interfaces
    * The database tables and proposed split of having some in one component and some in the other are an implementation detail
* If persistence details are encapsulated, then you cannot join across tables behind the boundary
* If you drop ItemTermStore and PropertyTermStore and just have wbt_item_terms/wbt_property_terms bound implementations in Wikibase Repo, then you loose the current implementation decoupling
* Having a service/component that just deals with terms and not the entity ids can make sense
* Such a component can be in its own sub domain. The code is already going in that direction by using data structures specific to the functionality provided. In this case it is likely good to explicitly not use Wikibase DataModel.

Replacing the current term store abstractions with the package private ones here makes little sense to me. Yes, we might want to reuse a bunch of the package private stuff here at some later point. We can then do the extraction and have an *additional* (lower level) public abstraction. I do not think it makes sense to do this now because of YAGNI and certainly don't think it makes sense to replace the current abstraction with the package private one.

Reply:
>         ItemTermStore and PropertyTermStore are good interfaces (which is a bit "doh!", since that is what they where created for)
Yes they are. No problem with these interfaces, and they will continue to exist in their context (Wikibase extension probably).
We have one of two options proceeding with Wikibase DataModel design:
- Drop the general abstraction. Meaning we drop EntityId and EntityDocument interfaces from the public spaces of the DataModel. If that's the direction that we want to go in, we should then not do the split suggested in here probably.
- Keep the abstraction but take Item and Property stuff out of it.
>         The database tables and proposed split of having some in one component and some in the other are an implementation detail
Not sure I understand this one.
>     If persistence details are encapsulated, then you cannot join across tables behind the boundary
No body says that joining tables is the only solution that must be used between linking tables. For that case specifically, it sounds like an implementation detail imposed on design, which sometimes is acceptable if such detail is of great importance for the function of the system. But it is not immediately obvious that it is to me yet.

To give and example, if normalized term tables are joined by a library to search through them and returns back ids always, then that is queries against linking tables in a simple query such as:
```
$resultTermIds = TermLookup::searchFuzzilyInAllTypesAndLanguages( 'some text' );
$resultItemIds = doQuery(<<<SQL
select wbit_id from wbt_item_terms
where wbit_term_in_lang_id IN { convertToSqlList( $resultTermIds ) }
SQL;
);
```
is something that would porbably be acceptable and necessarily less performant that doing all of it as one query and one series of joins. There is even another benefit with such split, now you can cache $resultTermIds separately too ;)

>     If you drop ItemTermStore and PropertyTermStore and just have wbt_item_terms/wbt_property_terms bound implementations in Wikibase Repo, then you loose the current implementation decoupling
No sure what you mean here. Would you give an example maybe on this?
>     Such a component can be in its own sub domain. The code is already going in that direction by using data structures specific to the functionality provided. In this case it is likely good to explicitly not use Wikibase DataModel.
Agree. but what is the difference between a domain and a sub domain? are they all domains on their own right at the end, and it is just a matter of the distance from where you view things? What difference would it make to this decision if we say "TermsStore" must be a sub domain and not a domain on its own right, practically?
> Replacing the current term store abstractions with the package private ones here makes little sense to me. Yes, we might want to reuse a bunch of the package private stuff here at some later point. We can then do the extraction and have an additional (lower level) public abstraction. I do not think it makes sense to do this now because of YAGNI and certainly don't think it makes sense to replace the current abstraction with the package private one.
Agree to YAGNI part.

----

I tend to agree with your concerns .. but I'm a bit paralyzed why you think the above decision is in contradiction with your main message. I think what that decision tries to do is the same as what you have described as "Such a component can be in its own sub domain. The code is already going in that direction by using data structures specific to the functionality provided. In this case it is likely good to explicitly not use Wikibase DataModel."

I feel like we are in agreement so not sure why that decision doesn't makes sense to you.

Or am I getting you completely wrong?

Reply:
> No problem with these interfaces, and they will continue to exist in their context (Wikibase extension probably).

Why move them there? Are the reasons why we created the library no longer valid?

> We have one of two options proceeding with Wikibase DataModel design

I do not see a benefit in including redesign of Wikibase DataModel in this conversation. I certainly was not talking about it.

>> The database tables and proposed split of having some in one component and some in the other are an implementation detail

> Not sure I understand this one.

The fact that there are DB tables is an implementation detail to begin with, so anything that depends on that is as well.

> No body says that joining tables is the only solution that must be used

Yes, and no one said anyone said that. All I did was point out you can't do joins with that approach. I totally agree with the rest of what you wrote.

> not sure why that decision doesn't makes sense to you.

The two reasons are described in the last paragraph of my previous comment.

Reply:
> If persistence details are encapsulated, then you cannot join across tables behind the boundary

> The fact that there are DB tables is an implementation detail to begin with, so anything that depends on that is as well.

I don’t fully agree with this. The exact internal layout of the tables is an implementation detail, but how can the mere fact that there are DB tables remain internal? Surely, Wikibase (or any other user of the library) will have to inject a database connection into the implementation. (The library may offer multiple term store implementations, some of which wouldn’t use a database, but I assume it will be Wikibase who chooses which implementation is used.)

And if the general fact that a database is used is part of the interface, then a join across the boundary should also be possible. For example, MediaWiki core’s `ActorMigration` and `CommentStore` classes each have a `getJoin()` method to instruct both other core classes and extensions how to perform a join against the actor/comment storage.

Reply:
Constructors are not part of the interface of a class. It does not matter what needs to be injected.